File: tests/network_operations/dns_lookup/dns_lookup.py
#!/usr/bin/env python3
"""
DNS lookup performance test implementation in Python.
Measures DNS resolution performance for various domain types.
"""

import json
import sys
import time
import socket
import concurrent.futures
from typing import Dict, List, Any
import threading


def resolve_domain(domain: str, timeout: float = 5.0) -> Dict[str, Any]:
    """Resolve a single domain and measure timing."""
    start_time = time.time()
    result = {
        "domain": domain,
        "success": False,
        "response_time_ms": 0.0,
        "ip_addresses": [],
        "error": None
    }
    
    try:
        # Resolve domain to IP addresses
        ip_addresses = socket.gethostbyname_ex(domain)[2]
        end_time = time.time()
        
        result.update({
            "success": True,
            "response_time_ms": (end_time - start_time) * 1000,
            "ip_addresses": ip_addresses
        })
        
    except socket.gaierror as e:
        end_time = time.time()
        result.update({
            "response_time_ms": (end_time - start_time) * 1000,
            "error": f"DNS resolution failed: {str(e)}"
        })
    except Exception as e:
        end_time = time.time()
        result.update({
            "response_time_ms": (end_time - start_time) * 1000,
            "error": f"Unexpected error: {str(e)}"
        })
    
    return result


def resolve_domains_sequential(domains: List[str], timeout: float = 5.0) -> List[Dict[str, Any]]:
    """Resolve domains sequentially."""
    results = []
    for domain in domains:
        result = resolve_domain(domain, timeout)
        results.append(result)
        print(f"  Resolved {domain}: {'✓' if result['success'] else '✗'} ({result['response_time_ms']:.2f}ms)", file=sys.stderr)
    return results


def resolve_domains_concurrent(domains: List[str], max_workers: int = 5, timeout: float = 5.0) -> List[Dict[str, Any]]:
    """Resolve domains concurrently using thread pool."""
    results = []
    
    with concurrent.futures.ThreadPoolExecutor(max_workers=max_workers) as executor:
        # Submit all tasks
        future_to_domain = {
            executor.submit(resolve_domain, domain, timeout): domain 
            for domain in domains
        }
        
        # Collect results as they complete
        for future in concurrent.futures.as_completed(future_to_domain):
            domain = future_to_domain[future]
            try:
                result = future.result()
                results.append(result)
                print(f"  Resolved {domain}: {'✓' if result['success'] else '✗'} ({result['response_time_ms']:.2f}ms)", file=sys.stderr)
            except Exception as e:
                error_result = {
                    "domain": domain,
                    "success": False,
                    "response_time_ms": 0.0,
                    "ip_addresses": [],
                    "error": f"Future execution failed: {str(e)}"
                }
                results.append(error_result)
                print(f"  Resolved {domain}: ✗ (future failed)", file=sys.stderr)
    
    # Sort results by domain name to maintain consistent order
    results.sort(key=lambda x: x["domain"])
    return results


def run_dns_benchmark(config: Dict) -> Dict:
    """Run DNS lookup benchmark."""
    domains = config.get("domains", ["google.com", "github.com", "stackoverflow.com"])
    resolution_modes = config.get("resolution_modes", ["sequential"])
    iterations = config.get("iterations", 3)
    timeout = config.get("timeout_seconds", 5.0)
    concurrent_workers = config.get("concurrent_workers", 5)
    
    results = {
        "start_time": time.time(),
        "test_cases": [],
        "summary": {
            "total_domains": len(domains),
            "total_iterations": 0,
            "successful_resolutions": 0,
            "failed_resolutions": 0,
            "avg_resolution_time": 0.0,
            "fastest_resolution": float('inf'),
            "slowest_resolution": 0.0
        }
    }
    
    all_resolution_times = []
    
    for mode in resolution_modes:
        print(f"Testing DNS resolution mode: {mode}...", file=sys.stderr)
        
        test_case = {
            "resolution_mode": mode,
            "domains_count": len(domains),
            "iterations": [],
            "avg_resolution_time": 0.0,
            "success_rate": 0.0,
            "total_time": 0.0
        }
        
        mode_resolution_times = []
        mode_successful = 0
        mode_total = 0
        
        for i in range(iterations):
            print(f"  Iteration {i+1}/{iterations}...", file=sys.stderr)
            
            iteration_start = time.time()
            
            if mode == "sequential":
                domain_results = resolve_domains_sequential(domains, timeout)
            elif mode == "concurrent":
                domain_results = resolve_domains_concurrent(domains, concurrent_workers, timeout)
            else:
                print(f"Warning: Unknown resolution mode '{mode}', using sequential", file=sys.stderr)
                domain_results = resolve_domains_sequential(domains, timeout)
            
            iteration_end = time.time()
            iteration_total_time = (iteration_end - iteration_start) * 1000  # ms
            
            # Calculate iteration statistics
            iteration_successful = sum(1 for r in domain_results if r["success"])
            iteration_failed = len(domain_results) - iteration_successful
            iteration_avg_time = sum(r["response_time_ms"] for r in domain_results if r["success"])
            if iteration_successful > 0:
                iteration_avg_time /= iteration_successful
            
            # Collect timing data
            for result in domain_results:
                if result["success"]:
                    mode_resolution_times.append(result["response_time_ms"])
                    all_resolution_times.append(result["response_time_ms"])
            
            mode_successful += iteration_successful
            mode_total += len(domain_results)
            results["summary"]["total_iterations"] += 1
            
            iteration_result = {
                "iteration": i + 1,
                "total_time_ms": iteration_total_time,
                "domains_resolved": len(domain_results),
                "successful_resolutions": iteration_successful,
                "failed_resolutions": iteration_failed,
                "avg_resolution_time_ms": iteration_avg_time,
                "domain_results": domain_results
            }
            
            test_case["iterations"].append(iteration_result)
        
        # Calculate test case averages
        if mode_resolution_times:
            test_case["avg_resolution_time"] = sum(mode_resolution_times) / len(mode_resolution_times)
            test_case["fastest_resolution"] = min(mode_resolution_times)
            test_case["slowest_resolution"] = max(mode_resolution_times)
        
        test_case["success_rate"] = (mode_successful / mode_total) * 100 if mode_total > 0 else 0
        test_case["total_successful"] = mode_successful
        test_case["total_attempts"] = mode_total
        
        results["test_cases"].append(test_case)
    
    # Calculate overall summary
    results["summary"]["successful_resolutions"] = sum(1 for t in all_resolution_times)
    results["summary"]["failed_resolutions"] = (
        results["summary"]["total_iterations"] * len(domains) - results["summary"]["successful_resolutions"]
    )
    
    if all_resolution_times:
        results["summary"]["avg_resolution_time"] = sum(all_resolution_times) / len(all_resolution_times)
        results["summary"]["fastest_resolution"] = min(all_resolution_times)
        results["summary"]["slowest_resolution"] = max(all_resolution_times)
    else:
        results["summary"]["fastest_resolution"] = 0.0
    
    results["end_time"] = time.time()
    results["total_execution_time"] = results["end_time"] - results["start_time"]
    
    return results


def main():
    """Main entry point for DNS lookup test."""
    if len(sys.argv) < 2:
        print("Usage: python dns_lookup.py <config_file>", file=sys.stderr)
        sys.exit(1)
    
    config_file = sys.argv[1]
    
    try:
        with open(config_file, 'r') as f:
            config = json.load(f)
        
        parameters = config.get("parameters", {})
        results = run_dns_benchmark(parameters)
        
        print(json.dumps(results, indent=2))
        
    except FileNotFoundError:
        print(f"Error: Config file '{config_file}' not found", file=sys.stderr)
        sys.exit(1)
    except json.JSONDecodeError as e:
        print(f"Error: Invalid JSON in config file: {e}", file=sys.stderr)
        sys.exit(1)
    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: tests/network_operations/dns_lookup/dns_lookup.ts
#!/usr/bin/env ts-node

import * as dns from 'dns';
import * as fs from 'fs';
import { promisify } from 'util';

const lookupAsync = promisify(dns.lookup);

interface DomainResult {
    domain: string;
    success: boolean;
    response_time_ms: number;
    ip_addresses: string[];
    error?: string;
}

interface IterationResult {
    iteration: number;
    total_time_ms: number;
    domains_resolved: number;
    successful_resolutions: number;
    failed_resolutions: number;
    avg_resolution_time_ms: number;
    domain_results: DomainResult[];
}

interface TestCase {
    resolution_mode: string;
    domains_count: number;
    iterations: IterationResult[];
    avg_resolution_time: number;
    success_rate: number;
    total_time: number;
    fastest_resolution?: number;
    slowest_resolution?: number;
    total_successful: number;
    total_attempts: number;
}

interface Summary {
    total_domains: number;
    total_iterations: number;
    successful_resolutions: number;
    failed_resolutions: number;
    avg_resolution_time: number;
    fastest_resolution: number;
    slowest_resolution: number;
}

interface BenchmarkResult {
    start_time: number;
    end_time: number;
    total_execution_time: number;
    test_cases: TestCase[];
    summary: Summary;
}

function resolveDomain(domain: string, timeout: number = 5000): Promise<DomainResult> {
    return new Promise((resolve) => {
        // For now, we're not implementing a cache since requiring external packages may not be supported
        const start = Date.now();
        const result: DomainResult = {
            domain,
            success: false,
            response_time_ms: 0,
            ip_addresses: []
        };

        // Set up timeout
        const timeoutId = setTimeout(() => {
            const end = Date.now();
            result.response_time_ms = end - start;
            result.error = 'Timeout';
            resolve(result);
        }, timeout);

        // Perform DNS lookup
        lookupAsync(domain, { family: 4 }, (err, address) => {
            // Clear timeout since we got a response
            clearTimeout(timeoutId);
            
            const end = Date.now();
            result.response_time_ms = end - start;
            
            if (err) {
                result.error = (err as Error).message;
                resolve(result);
            } else {
                result.success = true;
                result.ip_addresses = [address];
                resolve(result);
            }
        });
    });
}

async function resolveDomainsSequential(domains: string[], timeout: number = 5000): Promise<DomainResult[]> {
    const results: DomainResult[] = [];
    
    for (const domain of domains) {
        const result = await resolveDomain(domain, timeout);
        results.push(result);
        console.error(`  Resolved ${domain}: ${result.success ? '✓' : '✗'} (${result.response_time_ms.toFixed(2)}ms)`);
    }
    
    return results;
}

async function resolveDomainsConcurrent(domains: string[], timeout: number = 5000): Promise<DomainResult[]> {
    const promises = domains.map(domain => resolveDomain(domain, timeout));
    const results = await Promise.all(promises);
    
    for (const result of results) {
        console.error(`  Resolved ${result.domain}: ${result.success ? '✓' : '✗'} (${result.response_time_ms.toFixed(2)}ms)`);
    }
    
    // Sort results by domain name to maintain consistent order
    results.sort((a, b) => a.domain.localeCompare(b.domain));
    return results;
}

async function runDnsBenchmark(config: any): Promise<BenchmarkResult> {
    const domains = config.domains || ['google.com', 'github.com', 'stackoverflow.com'];
    const resolutionModes = config.resolution_modes || ['sequential'];
    const iterations = config.iterations || 3;
    const timeout = (config.timeout_seconds || 5) * 1000; // Convert to ms

    const startTime = Date.now();
    const testCases: TestCase[] = [];
    const allResolutionTimes: number[] = [];
    let totalIterations = 0;

    for (const mode of resolutionModes) {
        console.error(`Testing DNS resolution mode: ${mode}...`);

        const testCase: TestCase = {
            resolution_mode: mode,
            domains_count: domains.length,
            iterations: [],
            avg_resolution_time: 0,
            success_rate: 0,
            total_time: 0,
            total_successful: 0,
            total_attempts: 0
        };

        const modeResolutionTimes: number[] = [];
        let modeSuccessful = 0;
        let modeTotal = 0;

        for (let i = 0; i < iterations; i++) {
            console.error(`  Iteration ${i + 1}/${iterations}...`);
            totalIterations++;

            const iterationStart = Date.now();

            let domainResults: DomainResult[];
            if (mode === 'sequential') {
                domainResults = await resolveDomainsSequential(domains, timeout);
            } else if (mode === 'concurrent') {
                domainResults = await resolveDomainsConcurrent(domains, timeout);
            } else {
                console.error(`Warning: Unknown resolution mode '${mode}', using sequential`);
                domainResults = await resolveDomainsSequential(domains, timeout);
            }

            const iterationEnd = Date.now();
            const iterationTotalTime = iterationEnd - iterationStart;

            // Calculate iteration statistics
            const iterationSuccessful = domainResults.filter(r => r.success).length;
            const iterationFailed = domainResults.length - iterationSuccessful;
            const successfulTimes = domainResults.filter(r => r.success).map(r => r.response_time_ms);
            const iterationAvgTime = successfulTimes.length > 0 
                ? successfulTimes.reduce((a, b) => a + b, 0) / successfulTimes.length 
                : 0;

            // Collect timing data
            modeResolutionTimes.push(...successfulTimes);
            allResolutionTimes.push(...successfulTimes);

            modeSuccessful += iterationSuccessful;
            modeTotal += domainResults.length;

            const iterationResult: IterationResult = {
                iteration: i + 1,
                total_time_ms: iterationTotalTime,
                domains_resolved: domainResults.length,
                successful_resolutions: iterationSuccessful,
                failed_resolutions: iterationFailed,
                avg_resolution_time_ms: iterationAvgTime,
                domain_results: domainResults
            };

            testCase.iterations.push(iterationResult);
        }

        // Calculate test case averages
        if (modeResolutionTimes.length > 0) {
            testCase.avg_resolution_time = modeResolutionTimes.reduce((a, b) => a + b, 0) / modeResolutionTimes.length;
            testCase.fastest_resolution = Math.min(...modeResolutionTimes);
            testCase.slowest_resolution = Math.max(...modeResolutionTimes);
        }

        testCase.success_rate = modeTotal > 0 ? (modeSuccessful / modeTotal) * 100 : 0;
        testCase.total_successful = modeSuccessful;
        testCase.total_attempts = modeTotal;

        testCases.push(testCase);
    }

    const endTime = Date.now();
    const totalExecutionTime = (endTime - startTime) / 1000; // Convert to seconds

    // Calculate overall summary
    const successfulResolutions = allResolutionTimes.length;
    const failedResolutions = (totalIterations * domains.length) - successfulResolutions;
    const avgResolutionTime = allResolutionTimes.length > 0 
        ? allResolutionTimes.reduce((a, b) => a + b, 0) / allResolutionTimes.length 
        : 0;
    const fastestResolution = allResolutionTimes.length > 0 ? Math.min(...allResolutionTimes) : 0;
    const slowestResolution = allResolutionTimes.length > 0 ? Math.max(...allResolutionTimes) : 0;

    return {
        start_time: startTime / 1000, // Convert to seconds
        end_time: endTime / 1000,
        total_execution_time: totalExecutionTime,
        test_cases: testCases,
        summary: {
            total_domains: domains.length,
            total_iterations: totalIterations,
            successful_resolutions: successfulResolutions,
            failed_resolutions: failedResolutions,
            avg_resolution_time: avgResolutionTime,
            fastest_resolution: fastestResolution,
            slowest_resolution: slowestResolution
        }
    };
}

async function main() {
    if (process.argv.length < 3) {
        console.error('Usage: ts-node dns_lookup.ts <config_file>');
        process.exit(1);
    }

    const configFile = process.argv[2];

    try {
        const configData = fs.readFileSync(configFile, 'utf8');
        const config = JSON.parse(configData);
        const parameters = config.parameters || {};
        
        const results = await runDnsBenchmark(parameters);
        console.log(JSON.stringify(results, null, 2));

    } catch (error) {
        if (error instanceof Error) {
            if (error.message.includes('ENOENT')) {
                console.error(`Error: Config file '${configFile}' not found`);
            } else if (error.message.includes('JSON')) {
                console.error(`Error: Invalid JSON in config file: ${error.message}`);
            } else {
                console.error(`Error: ${error.message}`);
            }
        } else {
            console.error(`Error: ${String(error)}`);
        }
        process.exit(1);
    }
}

main();